roadmap/roadmap.adoc

```diff
@@ -0,0 +1,158 @@
+
+Open questions for this protocol include
+
+* Which messages do we support?
+For example, is _move a child from position 2 to position 5 in the same list_ the same as _move a child from parent A to parent B_ the same message kind?
+
+* Is the protocol the same for client -> repository updates vs. repository -> client updates?
+That's especially relevant for (web-based) editors.
+
+* How strong do we couple the protocol to an implementation?
+Both https://en.wikipedia.org/wiki/Operational_transformation[operational transformation] and https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type[CRDTs] can be used to implement consistent clients.
+For both of them, several implementations are available.
+Additionally, we might support simpler clients that assume no conflicts or missing updates at all.
+
+==== Standard derived models
+We intend to support several common language (workbench) functionalities via <<{refarch}.adoc#derivedModel, derived models>>.
+For each functionality, LionWeb would define a _base language_.
+Any implementation might use these base languages directly, or define its own extension languages.
+
+As an example, LionWeb's _validation base language_ might define concepts for `Error`, `Warning`, and `Info`.
+My validator implementation would create instances of `Error` and `Info`, but also include a custom _compiler-based validation language_ that defines `GeneratorWarning` and `CompilerWarning` as specializations of `Warning`.
+A client that only knows about LionWeb's _validation base language_ can process both of the specialized warnings in a generic fashion.
+Another, compiler-aware client might display ``GeneratorWarning``s differently than ``CompilerWarning``s.
+
+_Base language_ candidates are typically common language functionalities that might be expensive to calculate.
+This includes:
+
+===== Validation
+Checking a model for consistency and other errors is one of the most common operations on models.
+It might also be expensive: if we wanted to assure globally unique names, we had to check all named things in our model on every name change.
+The basic structure of validation is very similar in most systems: any issue has a severity level, refers to a node, might relate to a feature, and provides an error message.
+
+===== Scoping
+Calculating the scope of visible targets at any given place is necessary for both parser-based clients (for name binding / linking) and projectional clients / IDEs (for auto-completion).
+Similar to validation, the baseic structure of scoping is very similar in most systems; this _base language_ would condense the common concepts.
+
+===== Typesystem
+We need to calculate the type of a node in several contexts, including validation (e.g. "do the types of these assigned nodes match?") and scoping (e.g. "which members does the thing on the left side of the dot have?").
+
+===== Commenting / Review notes
```

Comment:
I'm moving this to a separate section, as reviews _cannot_ be derived models -- we cannot reconstruct them from the original models.